lib/FrontendTool/FrontendTool.cpp

```diff
@@ -699,29 +700,59 @@ static bool performCompile(CompilerInstance &Instance,
     IRGenOpts.DebugInfoKind = IRGenDebugInfoKind::Normal;
     const ProcessCmdLine &CmdLine = ProcessCmdLine(opts.ImmediateArgv.begin(),
                                                    opts.ImmediateArgv.end());
-    Instance.setSILModule(std::move(SM));
+    Instance->setSILModule(std::move(SM));
 
     if (observer) {
-      observer->aboutToRunImmediately(Instance);
+      observer->aboutToRunImmediately(*Instance);
     }
 
     ReturnValue =
-      RunImmediately(Instance, CmdLine, IRGenOpts, Invocation.getSILOptions());
-    return false;
+      RunImmediately(*Instance, CmdLine, IRGenOpts, Invocation.getSILOptions());
+    return Context.hadError();
   }
 
   // FIXME: We shouldn't need to use the global context here, but
   // something is persisting across calls to performIRGeneration.
   auto &LLVMContext = getGlobalLLVMContext();
+  std::unique_ptr<llvm::Module> IRModule;
+  llvm::GlobalVariable *HashGlobal;
   if (PrimarySourceFile) {
-    performIRGeneration(IRGenOpts, *PrimarySourceFile, std::move(SM),
-                        opts.getSingleOutputFilename(), LLVMContext);
+    IRModule = performIRGeneration(IRGenOpts, *PrimarySourceFile, std::move(SM),
+                                   opts.getSingleOutputFilename(), LLVMContext,
+                                   0, &HashGlobal);
   } else {
-    performIRGeneration(IRGenOpts, Instance.getMainModule(), std::move(SM),
-                        opts.getSingleOutputFilename(), LLVMContext);
-  }
-
-  return false;
+    IRModule = performIRGeneration(IRGenOpts, Instance->getMainModule(),
+                                   std::move(SM),
+                                   opts.getSingleOutputFilename(), LLVMContext,
+                                   &HashGlobal);
+  }
+
+  // Just because we had an AST error it doesn't mean we can't performLLVM.
+  bool HadError = Instance->getASTContext().hadError();
+  
+  // If the AST Context has no errors but no IRModule is available,
+  // parallelIRGen happened correctly.
```

Comment:
This is also too subtle. Maybe we should push the parallel-vs.-not decision up to which `perfom…` function you call? @eeckstein, what do you think?